### ISSUE TITLE
Warnings & errors not recognized on some Windows terminals

### ISSUE THREAD

**Issue by ultrapoci** [Canop/bacon#70]
I'm using Bacon 2.0.0 on Windows. If I run `bacon check` or `bacon clippy` and the program has errors, everything works as intended (even tho the shortcuts for wrapping lines and summary mode are not available, not sure if it is intended when there are errors), and the output looks something like this:
![image](https://user-images.githubusercontent.com/50493113/154506685-8cb56d8a-9cdd-4371-9c3c-2fc63b4285cd.png)

Instead, if only warnings are present, I can see bacon listing them for a fraction of a second, but then the screen turns to this:
![image](https://user-images.githubusercontent.com/50493113/154507070-70153c91-c059-468c-9adb-bcc484e8141f.png)

https://user-images.githubusercontent.com/50493113/154508472-6a549e44-edeb-4fc2-a184-ffa332a42464.mp4

Basically, only the number of warnings is showed and not the actual warnings themselves, and I hope you can see from the video that for a moment a longer output is showed. I've tested this on a newly created project. 

Rust version is 1.58.1.

**Comment by ultrapoci** [comment]
Output log of `bacon clippy`:
```16:10:41.684 [INFO] cli_log::init: Starting bacon v2.0.0 with log level DEBUG
16:10:41.686 [INFO] bacon::cli: args: Args {
    prefs: false,
    summary: false,
    no_summary: false,
    wrap: false,
    no_wrap: false,
    reverse: false,
    no_reverse: false,
    list_jobs: false,
    init: false,
    job: Some(
        "clippy",
    ),
    no_default_features: false,
    all_features: false,
    features: None,
    path: None,
    args: [],
    additional_job_args: [],
}
16:10:42.815 [DEBUG] bacon::cli: cargo_toml_file: "E:\\Programmazione\\rust\\bacon_test\\Cargo.toml"
16:10:42.828 [DEBUG] bacon::mission: add watch file "E:\\Programmazione\\rust\\bacon_test\\Cargo.toml"
16:10:42.830 [DEBUG] bacon::mission: add watch dir "E:\\Programmazione\\rust\\bacon_test\\src"
16:10:42.831 [DEBUG] bacon::mission: command: "cargo" "clippy" "--color" "always"
16:10:42.839 [DEBUG] bacon::executor: starting task Task { backtrace: false }
16:10:44.230 [DEBUG] bacon::executor: exit_status: ExitStatus(ExitStatus(0))
16:10:44.230 [DEBUG] bacon::executor: finished command execution
16:10:44.230 [INFO] bacon::app: execution finished with status: Some(ExitStatus(ExitStatus(0)))
16:10:44.230 [DEBUG] bacon::report: cmd_line=CommandOutputLine { content: TLine { strings: [TString { csi: "\u{1b}[1m\u{1b}[38;5;11m", raw: "warning" }, TString { csi: "\u{1b}[1m\u{1b}[38;5;15m", raw: ": unused variable: `asd`" }] }, origin: StdErr }
16:10:44.231 [DEBUG] bacon::report: line_analysis=LineAnalysis { line_type: Normal, key: None }
16:10:44.231 [DEBUG] bacon::report: cmd_line=CommandOutputLine { content: TLine { strings: [TString { csi: "", raw: " " }, TString { csi: "\u{1b}[1m\u{1b}[38;5;14m", raw: "--> " }, TString { csi: "", raw: "src\\main.rs:2:9" }] }, origin: StdErr }
16:10:44.231 [DEBUG] bacon::report: line_analysis=LineAnalysis { line_type: Normal, key: None }
16:10:44.231 [DEBUG] bacon::report: cmd_line=CommandOutputLine { content: TLine { strings: [TString { csi: "", raw: "  " }, TString { csi: "\u{1b}[1m\u{1b}[38;5;14m", raw: "|" }] }, origin: StdErr }
16:10:44.232 [DEBUG] bacon::report: line_analysis=LineAnalysis { line_type: Normal, key: None }
16:10:44.232 [DEBUG] bacon::report: cmd_line=CommandOutputLine { content: TLine { strings: [TString { csi: "\u{1b}[1m\u{1b}[38;5;14m", raw: "2" }, TString { csi: "", raw: " " }, TString { csi: "\u{1b}[1m\u{1b}[38;5;14m", raw: "| " }, TString { csi: "", raw: "    let asd = 13;" }] }, origin: StdErr }
16:10:44.232 [DEBUG] bacon::report: line_analysis=LineAnalysis { line_type: Normal, key: None }
16:10:44.232 [DEBUG] bacon::report: cmd_line=CommandOutputLine { content: TLine { strings: [TString { csi: "", raw: "  " }, TString { csi: "\u{1b}[1m\u{1b}[38;5;14m", raw: "| " }, TString { csi: "", raw: "        " }, TString { csi: "\u{1b}[1m\u{1b}[38;5;11m", raw: "^^^" }, TString { csi: "", raw: " " }, TString { csi: "\u{1b}[1m\u{1b}[38;5;11m", raw: "help: if this is intentional, prefix it with an underscore: `_asd`" }] }, origin: StdErr }
16:10:44.233 [DEBUG] bacon::report: line_analysis=LineAnalysis { line_type: Normal, key: None }
16:10:44.233 [DEBUG] bacon::report: cmd_line=CommandOutputLine { content: TLine { strings: [TString { csi: "", raw: "  " }, TString { csi: "\u{1b}[1m\u{1b}[38;5;14m", raw: "|" }] }, origin: StdErr }
16:10:44.233 [DEBUG] bacon::report: line_analysis=LineAnalysis { line_type: Normal, key: None }
16:10:44.233 [DEBUG] bacon::report: cmd_line=CommandOutputLine { content: TLine { strings: [TString { csi: "", raw: "  " }, TString { csi: "\u{1b}[1m\u{1b}[38;5;14m", raw: "= " }, TString { csi: "\u{1b}[1m\u{1b}[38;5;15m", raw: "note" }, TString { csi: "", raw: ": `#[warn(unused_variables)]` on by default" }] }, origin: StdErr }
16:10:44.234 [DEBUG] bacon::report: line_analysis=LineAnalysis { line_type: Normal, key: None }
16:10:44.234 [DEBUG] bacon::report: cmd_line=CommandOutputLine { content: TLine { strings: [] }, origin: StdErr }
16:10:44.234 [DEBUG] bacon::report: line_analysis=LineAnalysis { line_type: Normal, key: None }
16:10:44.234 [DEBUG] bacon::report: cmd_line=CommandOutputLine { content: TLine { strings: [TString { csi: "\u{1b}[1m\u{1b}[33m", raw: "warning" }, TString { csi: "\u{1b}[1m", raw: ":" }, TString { csi: "", raw: " `bacon_test` (bin \"bacon_test\") generated 1 warning" }] }, origin: StdErr }
16:10:44.235 [DEBUG] bacon::report: line_analysis=LineAnalysis { line_type: Title(Warning), key: None }
16:10:44.235 [DEBUG] bacon::report: cmd_line=CommandOutputLine { content: TLine { strings: [TString { csi: "\u{1b}[1m\u{1b}[32m", raw: "    Finished" }, TString { csi: "", raw: " dev [unoptimized + debuginfo] target(s) in 0.39s" }] }, origin: StdErr }
16:10:44.235 [DEBUG] bacon::report: line_analysis=LineAnalysis { line_type: Normal, key: None }
16:10:44.235 [DEBUG] bacon::state: GOT REPORT
16:10:50.357 [DEBUG] bacon::app: key pressed: q
16:10:50.357 [DEBUG] bacon::app: requested action: Internal(Quit)
16:10:50.357 [DEBUG] bacon::executor: received kill order
16:10:50.357 [DEBUG] bacon::executor: leaving thread
16:10:50.358 [INFO] bacon: bye
```

Here is a MEGA link with the project: https://mega.nz/file/X9wknS7a#97wLzZ7k3U8KXGU79Pgm5sPAVKGCq6iUVQwp57HuGjI
As I've said, it is basically a newly created project with `cargo new`.

**Comment by Canop** [comment]
I've ran `bacon clippy` in your project and got this:

![image](https://user-images.githubusercontent.com/617006/154512936-c58adb39-4adb-4bdc-94a9-9e7919f7b162.png)

The bug might be related to the terminal you use. Which one is it ? Do you have other terminals available ?

**Comment by Canop** [comment]
Could you try with this bacon.toml:

```
# This is a configuration file for the bacon tool
# More info at https://github.com/Canop/bacon

default_job = "check"

[jobs]

[jobs.check]
command = ["cargo", "check", "--color", "always"]
need_stdout = false

[jobs.check-all]
command = ["cargo", "check", "--all-targets", "--color", "always"]
need_stdout = false
watch = ["tests", "benches", "examples"]

[jobs.clippy]
command = ["cargo", "clippy", "--color", "always"]
need_stdout = false

[jobs.clippy-all]
command = ["cargo", "clippy", "--all-targets", "--color", "always"]
need_stdout = false
watch = ["tests", "benches", "examples"]

[jobs.test]
command = ["cargo", "test", "--color", "always"]
need_stdout = true
watch = ["tests"]

[jobs.doc]
command = ["cargo", "doc", "--color", "always", "--no-deps"]
need_stdout = false

# if the doc compiles, then it opens in your browser and bacon switches
# to the previous job
[jobs.doc-open]
command = ["cargo", "doc", "--color", "always", "--no-deps", "--open"]
need_stdout = false
on_success = "back" # so that we don't open the browser at each change

# You can run your application and have the result displayed in bacon,
# *if* it makes sense for this crate. You can run an example the same
# way. Don't forget the `--color always` part or the errors won't be
# properly parsed.
[jobs.run]
command = ["cargo", "run", "--color", "always"]
need_stdout = true

# You may define here keybindings that would be specific to
# a project, for example a shortcut to launch a specific job.
# Shortcuts to internal functions (scrolling, toggling, etc.)
# should go in your personal prefs.toml file instead.
[keybindings]
a = "job:check-all"
i = "job:initial"
c = "job:clippy"
d = "job:doc-open"
t = "job:test"
r = "job:run"
o = "toggle-raw-output"
```

(it adds the <kbd>o</kbd> shortcut to `toggle-raw-output`), launch `bacon clippy`, then hit the <kbd>o</kbd> key and screenshot the result ?

**Comment by ultrapoci** [comment]
I'm using the Windows Terminal, but using the other terminal installed (cmd) doesn't change anything.

Using that bacon.toml and pressing `o` gives this:
![image](https://user-images.githubusercontent.com/50493113/154516467-9d0c9d6b-d768-4b65-aca8-e165cbdd150d.png)

If an error is present, the output is the same as before, and the `o` shortcut doesn't do anythin:
![image](https://user-images.githubusercontent.com/50493113/154516758-27c5a754-ad31-48f2-9771-4b233071431c.png)

**Comment by julienc** [comment]
I have this issue as well on Windows, in both powershell and plain cmd.exe.

**Comment by julienc** [comment]
Update: I just upgraded to version 2.0.1 and the issue appears to be fixed.

**Comment by Canop** [comment]
> Update: I just upgraded to version 2.0.1 and the issue appears to be fixed.

This is very unexpected

**Comment by julienc** [comment]
In 2.0, installed with cargo install.

**Comment by ultrapoci** [comment]
I've updated to 2.0.1. Here are some screenshots:
- one warning 
![image](https://user-images.githubusercontent.com/50493113/154761288-e9bb9fc4-8cf8-430a-b351-fac7dc2f1199.png)

- one error 
![image](https://user-images.githubusercontent.com/50493113/154761303-06c80b10-e530-4d59-b882-c523f3ad2b45.png)

- warning and error 
![image](https://user-images.githubusercontent.com/50493113/154761315-16dde374-a669-402a-bd55-aa3d73abf095.png)

Overall it works as intended, but there's no number in front of warnings and errors (which were present before). Also, the `s` and `w` shortcuts don't seem to do anything, but maybe it is due to the output not showing any difference (even tho in 2.0.0 the `s` shortcut used to hide the line about the building process being finished, but now that line is always shown despite having pressed `s`). Note that other shortcuts, such as `h` or `r`, do work.

The project used is the same as before (a simple hello world program), but I've removed the `bacon.toml` file before testing.

**Comment by Canop** [comment]
> Overall it works as intended

No: parsing doesn't seem to work at all. It looks like some styling information, which is used for parsing, is lost.

**Comment by Elarcis** [comment]
I have this issue as well, I’m using VSCode’s integrated terminal.

It only seems to arise when using Powershell though, as if I start a WSL terminal, I get proper warnings.

Powershell is on the left, Debian (WSL) on the right:
![image](https://user-images.githubusercontent.com/990841/179428467-0fe72b64-894c-47b0-91ac-746fe158ac23.png)

Trying to enable summary mode gives this:
![image](https://user-images.githubusercontent.com/990841/179428496-560ba267-2b5e-4922-8b97-a33cec357a8a.png)

**Comment by ethanniser** [comment]
![image](https://user-images.githubusercontent.com/100045248/216729033-f9644fa9-fd73-4285-8bec-27bb7c55bc45.png)
![image](https://user-images.githubusercontent.com/100045248/216729127-478e9637-c29e-4822-bc38-e89812e4ad67.png)
I just installed Bacon for the first time and I'm having this issue. Works fine in run/test mode, but in clippy mode the warnings don't get spelled out.

**Comment by ethanniser** [comment]
Still can't figure this out. The warnings show as expected but only for a flash then I get the collapsed message
![](https://user-images.githubusercontent.com/100045248/217112459-423a788f-f094-42ef-b18d-93835b834ce8.gif)

**Comment by Canop** [comment]
I don't have any Windows computer. Somebody else than me should try to decipher the problem and see whether it's easy to fix.
If you need help making sense of bacon't logic and code, you may come to my chat: https://miaou.dystroy.org/3768

**Comment by ethanniser** [comment]
Admittedly I am still learning rust and definitely unfamiliar with this code base, but I will try my best and take a look when I get home tonight.

**Comment by sylbeth** [comment]
Seems like the bug is persistent, and now I can't downgrade to 2.0.1 because of some issues back there:

![image](https://user-images.githubusercontent.com/102521501/220813412-bc02cb77-3fd2-42da-974b-d3524a14c80e.png)


This is what my terminal shows:

https://user-images.githubusercontent.com/102521501/220811774-c404949e-2c75-4469-91dc-55a4f4c6416e.png

vs what it should show

https://user-images.githubusercontent.com/102521501/220811831-e64ca5f7-485b-4f4a-8ed5-f05bbeb48f04.png

**Comment by ethanniser** [comment]
Sorry I never provided an update. The most I was able to gather was for some reason whenever I would put in a unused env key, that warning would be added at the top of the output, and all of the warnings would appear. 
![Screenshot 2023-02-07 184837](https://user-images.githubusercontent.com/100045248/220814160-36f88ed9-bcab-4cc6-b214-b079db6633e5.png)
But without the unused env the error would occur and the warnings wouldn't show.
![Screenshot 2023-02-07 184513](https://user-images.githubusercontent.com/100045248/220814181-8ab630cd-f703-455a-9970-415edf87640d.png)

Not sure how useful this is, but it was interesting when I stumbled into it.

Best of luck to whoever picks this up next.

**Comment by crillon** [comment]
@ultrapoci @Elarcis @sylbeth @Elarcis 
Hello, i made a pull request #120 with a fix. 
But it's better to test it on other Windows machines before approval.
Can any of you test it?

To install bacon with a fix:
```
$ cargo uninstall bacon
$ cargo install --git https://github.com/crillon/bacon.git
```

To revert:
```
$ cargo uninstall bacon
$ cargo install bacon
```
![image](https://user-images.githubusercontent.com/43761130/224687612-f8847a1f-2b39-4fbc-a495-bb4474ddc1fc.png)
